pkg/operation/botanist/vpnseedserver.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package botanist
+
+import (
+	"context"
+
+	"github.com/gardener/gardener/charts"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/operation/botanist/component"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/vpnseedserver"
+	"github.com/gardener/gardener/pkg/utils"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+
+	"k8s.io/utils/pointer"
+)
+
+const diffieHellmanKey = "LS0tLS1CRUdJTiBESCBQQVJBTUVURVJTLS0tLS0KTUlJQkNBS0NBUUVBN2NCWHhHOWFuNktSei9zQjV1aVNPVGY3RWcrdVdWa2hYTzRwZUtEVEFSek1ZYThiN1dSOApCL0F3K0F5VVh0QjN0WHRyemVDNU0zSUhudWhGd01vM0s0b1NPa0ZKeGF0TGxZS2VZMTVyK0t0NXZuT09UM0JXCmVONU9uV2xSNVdpN0daQldiYVFnWFZSNzlONHlzdDQzc1ZoSnVzNkJ5MGxONk9sYzl4RC95czlHSC95a0pWSWgKWi9OTHJ4QUM1bHhqd0NxSk1kOGhycnlDaHVEbHo1OTd2ZzZnWUZ1UlY2MFUvWVU0REs3MUY0SDdtSTA3YUdKOQpsK1NLOFRia0tXRjVJVEk3a1lXYmM0em10ZlhTWGFHak1oTTlvbVFVYVRIOWNzQjk2aHpGSmRlWjRYanh5YlJmClZjM3Q3WFA1cTdhZmVhS21NM0ZoU1hkZUhLQ1RxUXpRdXdJQkFnPT0KLS0tLS1FTkQgREggUEFSQU1FVEVSUy0tLS0tCg=="
+
+// DefaultVPNSeedServer returns a deployer for the vpn-seed-server.
+func (b *Botanist) DefaultVPNSeedServer() (vpnseedserver.VPNSeedServer, error) {
+	imageAPIServerProxy, err := b.ImageVector.FindImage(charts.ImageNameApiserverProxy, imagevector.RuntimeVersion(b.SeedVersion()), imagevector.TargetVersion(b.ShootVersion()))
+	if err != nil {
+		return nil, err
+	}
+
+	imageVPNSeedServer, err := b.ImageVector.FindImage(charts.ImageNameVpnSeedServer, imagevector.RuntimeVersion(b.SeedVersion()), imagevector.TargetVersion(b.ShootVersion()))
+	if err != nil {
+		return nil, err
+	}
+
+	var kubeAPIServerHost *string
+	if b.APIServerSNIEnabled() {
+		kubeAPIServerHost = pointer.StringPtr(b.outOfClusterAPIServerFQDN())
+	}
+
+	return vpnseedserver.New(
+		b.K8sSeedClient.Client(),
+		b.Shoot.SeedNamespace,
+		imageAPIServerProxy.String(),
+		imageVPNSeedServer.String(),
+		kubeAPIServerHost,
+		b.Shoot.Networks.Services.String(),
+		b.Shoot.Networks.Pods.String(),
+		b.Shoot.GetNodeNetwork(),
+		b.Shoot.GetReplicas(1),
+	), nil
+}
+
+func (b *Botanist) DeployVPNServer(ctx context.Context) error {
+
+	if b.Shoot.KonnectivityTunnelEnabled || !b.Shoot.ReversedVPNEnabled {
```

Comment:
Why do you care about the `KonnectivityTunnelEnabled` setting here? Shouldn't both be mutually exclusive? This could be validated early (start-up of gardenlet) so that you don't need to handle it here.

Reply:
Just like konnectivity tunnel, reverse vpn tunnel has an annotation that overrides the feature gate per cluster and allows to enable the feature for a cluster even if the feature gate is disabled. This is why we need to check at reconciliation whether the feature set is sensible.
Do you have another idea how to handle it?

Reply:
Hm, I didn't understand it fully, can you explain again? Particularly, why is `DeployVPNSeedServer` in any way depending on the `KonnectivityTunnel`? Yes, if the tunnel is enabled then it should be uninstalled, of course. What I was proposing was to ensure that in this case (tunnel enabled) the reversed VPN feature gate must be disabled. If this would be ensured then it would be enough to only keep 

```suggestion
	if !b.Shoot.ReversedVPNEnabled {
```

right?

Reply:
We addressed it in the new commit.